### PR TITLE
gh-112144: Make `http.cookiejar.CookieJar` picklable

### LINKED ISSUE #112144: Make `http.cookiejar.CookieJar` picklable
# Bug report

### Bug description:

I am developing a program that employs multiprocessing to enhance efficiency through parallel execution. Python relies on the `pickle` module for object serialization and inter-process communication. However, I've encountered an issue where the `CookieJar` object cannot be pickled due to its inclusion of an RLock object, which is causing problems in my application.

Here is the membership hierarchy:

```text
openai.OpenAI -> httpx.Client -> httpx.Cookies -> http.cookiejar.CookieJar
```

Reproduce snippet:

```python
# Add a code block here, if required

import pickle
from http.cookiejar import CookieJar


jar = CookieJar()
pickle.dumps(jar)  # TypeError: cannot pickle '_thread.RLock' object
```

We could implement a `__reduce__` method for the `CookieJar` class to let it support serialization.

### CPython versions tested on:

3.11

### Operating systems tested on:

macOS

### Linked PRs
* gh-112178

### PR DESCRIPTION
**Pull request by XuehaiPan**
This PR adds `__getstate__` and `__setstate__` methods to `http.cookiejar.CookieJar` to make it picklable. The `__getstate__` method gets all attributes of the `CookieJar` object except the `RLock`.

Resolves #112144

* Issue: gh-112144

### CODE REVIEW

**Review by terryjreedy — CHANGES_REQUESTED** [review]
I don't know anything about why this lock is present and whether anything needs doing when unpickling; I leave that to another reviewer.

This new feature should have a `test_pickle` added to test/test_http_cookiejar.py.  Since there is no CookiejarTest class, FileCookieJar seems like the most likely place for it.  At least check that round-tripping, unpickle(pickle(jar)) == jar, with any needed adjustment for the lock entry, works as expected.

**Comment by XuehaiPan** [comment]
I have made the requested changes; please review again

**Pull request closed without merging**
